keps/sig-apps/3017-pod-healthy-policy-for-pdb/README.md

```diff
@@ -0,0 +1,429 @@
+in a way that doesn't support this use-case is not an option. In particular, as Kubernetes
+doesn't provide any alternatives solutions for this problem.
+
+### Goals
+
+- Prevent PDBs from deadlocking eviction due to non-Ready pods.
+- Make sure users who rely on PDBs for data-safety can continue to do so.
+
+### Non-Goals
+
+- Providing a safe solution for preventing data-loss. Not because this isn't important, but 
+it is unclear if PDB is the right tool for this.
+
+## Proposal
+
+The core issue here is whether a pod that is Running but not Ready is already disrupted, 
+and thus can be evicted without being constrained by a potential Pod Disruption Budget. Currently,
+the disruption controller only considers Running and Ready pods as healthy, thus that is 
+the basis for computing `allowedDisruptions`. The disruption API on the other hand, does require
+that `allowedDisruptions` is larger than 0 to evict a pod that is Running but not Ready. 
+
+Adding a `podHealthyPolicy` field on the PDB API will allow the user to specify which
+behavior that are desired, and this will be consistently handled by the disruption controller, eviction
+API, and any other APIs that might use PDBs. If a `podHealthyPolicy` is not provided, the default
+will be the current behavior.
+
+### Risks and Mitigations
+
+## Design Details
+
+### API
+
+```golang
+// PodDisruptionBudgetSpec is a description of a PodDisruptionBudget.
+type PodDisruptionBudgetSpec struct {
+	
+	...
+	
+	// PodHealthyPolicy defines the criteria for when the disruption controller
+	// should consider a pod to be healthy.
+	// If no policy is specified, the legacy behavior will be used. It means
+	// only pods that are Running and Ready will be considered when the disruption
+	// controller computes "disruptionsAllowed", but all pods in the Running phase
+	// will be subject to the PDB on eviction.
+	// +optional
+	PodHealthyPolicy PodHealthyPolicy `json:"podHealthyPolicy,omitempty" protobuf:"bytes,4,opt,name=podHealthyPolicy"`
+}
+
+// PodHealthyPolicy defines the policy when a pod are considered healthy and therefore
+// covered by a PodDisruptionBudget.
+type PodHealthyPolicy string
+
+const (
+	// PodReady policy means that only pods that are both Running and Ready
+	// will be considered healthy by the disruption controller. Any pods that
+	// are not Ready are considered to already be disrupted and therefore will
+	// not be counted when computing "disruptionsAllowed" and can be evicted
+	// regardless of whether the criteria in a PDB is met.
+	PodReady PodHealthyPolicy = "PodReady"
+	
+	// PodRunning policy means that pods that are in the Running phase
+	// is considered healthy by the disruption controller, regardless of
+	// whether they are Ready or not. Any pods that are in the Running
+	// phase will be counted when computing "disruptionsAllowed" and
+	// will be subject to the PDB for eviction.
+	PodRunning PodHealthyPolicy = "PodRunning"
```

Comment:
That's the current default, maybe add an info that empty or this means the same.

Reply:
Currently the disruption controller and the eviction api don't agree on what is needed for a pod to be healthy, so we don't have a policy here that reflect the current behavior. The idea is that by specifying a policy, users can get consistent behavior between the two.